File: Aran/Geometries/Base/trunk/Spatial Reference/SpatialReference.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Aran;
using Aran.Package;

namespace Aran.Geometries.SpatialReferences
{
	public class SpatialReference : IPackable
	{
		public SpatialReference()
		{
			_ellipsoid = new Ellipsoid();
			_paramList = new AranObjectList<SpatialReferenceParam>();
			SpatialReferenceType = SpatialReferenceType.srtGeographic;
			SpatialReferenceUnit = SpatialReferenceUnit.sruMeter;
		}

		public Ellipsoid Ellipsoid
		{
			get
			{
				return _ellipsoid;
			}
		}

		public AranObjectList<SpatialReferenceParam> ParamList
		{
			get
			{
				return _paramList;
			}
		}

		public double this[SpatialReferenceParamType param]
		{

			get
			{
				SpatialReferenceParam PandaItem;
				for (int i = 0; i < ParamList.Count; i++)
				{
					PandaItem = ParamList[i];
					if (PandaItem.SRParamType == param)
						return PandaItem.Value;
				}
				return 0;
			}

			set
			{
				SpatialReferenceParam PandaItem;
				for (int i = 0; i < ParamList.Count; i++)
				{
					PandaItem = ParamList[i];
					if (PandaItem.SRParamType == param)
						PandaItem.Value = value;
				}
			}
		}

		public string Name { get; set; }

		public SpatialReferenceType SpatialReferenceType { get; set; }

		public SpatialReferenceUnit SpatialReferenceUnit { get; set; }

		public void Pack(PackageWriter writer)
		{
			writer.PutString(Name);
			writer.PutEnum<SpatialReferenceType>(SpatialReferenceType);
			writer.PutEnum<SpatialReferenceUnit>(SpatialReferenceUnit);
			Ellipsoid.Pack(writer);

			writer.PutInt32(_paramList.Count);
			foreach (SpatialReferenceParam srParam in _paramList)
				srParam.Pack(writer);
		}

		public void Unpack(PackageReader reader)
		{
			Name = reader.GetString();
			SpatialReferenceType = reader.GetEnum<SpatialReferenceType>();
			SpatialReferenceUnit = reader.GetEnum<SpatialReferenceUnit>();
			Ellipsoid.Unpack(reader);

			int count = reader.GetInt32();
			for (int i = 0; i < count; i++)
			{
				SpatialReferenceParam srParam = new SpatialReferenceParam();
				srParam.Unpack(reader);
				_paramList.Add(srParam);
			}
		}

		private Ellipsoid _ellipsoid;
		private AranObjectList<SpatialReferenceParam> _paramList;
	}
}

File: Aran/RNAV/Libs/Panda/GeometryOperatorsContract/cpp/src/SpatialReference.cpp
#include "../include/SpatialReference.h"

namespace Panda
{
	SpatialReferenceParam::SpatialReferenceParam() :
		srParamType(0),
		value(0)
	{
	}

	SpatialReferenceParam::SpatialReferenceParam(int SRParamType, double Value)
	{
		srParamType = SRParamType;
		value = Value;
	}

	SpatialReferenceParam::~SpatialReferenceParam()
	{
	}

	void SpatialReferenceParam::pack (Handle handle) const throw (Registry::Exception)
	{
		Registry::putInt32 (handle, srParamType);
		Registry::putDouble(handle, value);
	}

	void SpatialReferenceParam::unpack (Handle handle) throw (Registry::Exception)
	{
		Registry::getInt32 (handle, srParamType);
		Registry::getDouble(handle, value);
	}

	void SpatialReferenceParam::assign (const SpatialReferenceParam& src)
	{
		srParamType = src.srParamType;
		value = src.value;
	}

	SpatialReferenceParam* SpatialReferenceParam::clone () const
	{
		SpatialReferenceParam* val = new SpatialReferenceParam();
		val->assign (*this);
		return val;
	}

//--------------------------------------------------------------------
//--------------------------------------------------------------------
//--------------------------------------------------------------------

	SpatialReference::SpatialReference() :
		name(L""),
		spatialReferenceType(0),
		spatialReferenceUnit(0)
	{
	}

	SpatialReference::~SpatialReference()
	{
	}

	void SpatialReference::pack (Handle handle) const throw (Registry::Exception)
	{
		Registry::putStdString (handle, name);
		Registry::putInt32 (handle, spatialReferenceType);
		Registry::putInt32 (handle, spatialReferenceUnit);
		ellipsoid.pack (handle);
		paramList.pack (handle);
	}

	void SpatialReference::unpack (Handle handle) throw (Registry::Exception)
	{
		Registry::getStdString (handle, name);
		spatialReferenceType = Registry::getInt32 (handle);
		Registry::getInt32 (handle, spatialReferenceUnit);
		ellipsoid.unpack (handle);
		paramList.unpack (handle);
	}
}